README.md
sbt-sonatype plugin
======

A sbt plugin for publishing your project to the Maven central repository through the REST API of Sonatype Nexus. Deploying artifacts to Sonatype repository is a requirement for synchronizing your projects to the [Maven central repository](https://repo1.maven.org/maven2/). __sbt-sonatype__ plugin enables two-step release of your Scala/Java projects.

 * `publishSigned` (with [sbt-pgp plugin](http://www.scala-sbt.org/sbt-pgp/))
    * Create GPG signed artifacts to a local staging repository.
    * Make sure adding `publishTo := sonatypePublishToBundle.value` to your build.sbt
 * `sonatypeBundleRelease` (New in sbt-sonatype 3.4)
    * This command will prepare a new remote staging repository at Sonatype. If there are exisiting staging repositories that have the same description with `sonatypeSessionName` key, sbt-sonatype will discard them properly.
    * Then, it will upload the artifacts in the local staging folder to the remote staging repository. Uploading artifacts as a bundle is much faster than uploading each artifact to Sonatype. For example, thousands of files can be uploaded in several minutes with bundle upload.
    * Finally, this command will perform the close and release steps at the Sonatype Nexus repository to verify the Maven central requirements.

 After these steps, your project will be synchronized to the Maven central within ten minutes. No longer need to enter the web interface of
 [Sonatype Nexus repository](http://oss.sonatype.org/) to performe these release steps.


- [Release notes](ReleaseNotes.md)
- sbt-sonatype is available for sbt 1.x series.
- You can also use sbt-sonatype for [publishing non-sbt projects](README.md#publishing-maven-projects) (e.g., Maven, Gradle, etc.)
- Blog: [Blazingly Fast Release to Sonatype](https://medium.com/@taroleo/sbt-sonatype-f02bdafd78f1)

## Prerequisites

 * Create a Sonatype Repository account
   * Follow the instruction in the [Central Repository documentation site](http://central.sonatype.org).
     * Create a Sonatype account
     * Create a GPG key
     * Open a JIRA ticket to get a permission for synchronizing your project to the Central Repository (aka Maven Central).


 * Related articles:
    * [Deploying to Sonatype - sbt Documentation](http://www.scala-sbt.org/release/docs/Community/Using-Sonatype.html)
    * [Uploading to a Staging Repository via REST API](https://support.sonatype.com/hc/en-us/articles/213465868-Uploading-to-a-Staging-Repository-via-REST-API)

## Configurations

[![Maven Central](https://maven-badges.herokuapp.com/maven-central/org.xerial.sbt/sbt-sonatype/badge.svg)](https://maven-badges.herokuapp.com/maven-central/org.xerial.sbt/sbt-sonatype)

### project/plugins.sbt

Import ***sbt-sonatype*** plugin and [sbt-pgp plugin](http://www.scala-sbt.org/sbt-pgp/) to use `sonatypeBundleRelease` and `publishSigned`
commands:
```scala
// For sbt 1.x (sbt-sonatype 2.3 or higher)
addSbtPlugin("org.xerial.sbt" % "sbt-sonatype" % "(version)")
addSbtPlugin("com.github.sbt" % "sbt-pgp" % "2.1.2")

// For sbt 0.13.x (upto sbt-sonatype 2.3)
addSbtPlugin("org.xerial.sbt" % "sbt-sonatype" % "(version)")
addSbtPlugin("com.jsuereth" % "sbt-pgp" % "1.0.0")
```

 * If downloading the sbt-sonatype plugin fails, check the repository in the Maven central: <https://repo1.maven.org/maven2/org/xerial/sbt/sbt-sonatype_2.12_1.0>. It will be usually synced within 10 minutes.

### build.sbt

To use sbt-sonatype, you need to create a bundle of your project artifacts (e.g., .jar, .javadoc, .asc files, etc.) into a local folder specified by `sonatypeBundleDirectory`. By default, the folder is `(project root)/target/sonatype-staging/(version)`. Add the following `publishTo` setting to create a local bundle of your project:
```scala
publishTo := sonatypePublishToBundle.value
```

  > ⚠️ Legacy Host
  >
  > By default, this plugin is configured to use the legacy Sonatype repository `oss.sonatype.org`. If you created a new account on or after February 2021, add `sonatypeCredentialHost` settings:
  >
  > ```scala
  > // For all Sonatype accounts created on or after February 2021
  > ThisBuild / sonatypeCredentialHost := "s01.oss.sonatype.org"
  > ```

With this setting, `publishSigned` will create a bundle of your project to the local staging folder. If the project has multiple modules, all of the artifacts will be assembled into the same folder to create a single bundle.

If `isSnapshot.value` is true (e.g., if the version name contains -SNAPSHOT), publishSigned task will upload files to the Sonatype Snapshots repository without using the local bundle folder.

If necessary, you can tweak several configurations:
```scala
// [Optional] The local staging folder name:
sonatypeBundleDirectory := (ThisBuild / baseDirectory).value / target.value.getName / "sonatype-staging" / (ThisBuild / version).value

// [Optional] If you need to manage unique session names by yourself, change this default setting:
sonatypeSessionName := s"[sbt-sonatype] ${name.value} ${version.value}"

// [Optional] Timeout until giving up sonatype close/promote stages. Default is 60 min.
sonatypeTimeoutMillis := 60 * 60 * 1000 

// [If you cannot use bundle upload] Use this setting when you need to uploads artifacts directly to Sonatype
// With this setting, you cannot use sonatypeBundleXXX commands
publishTo := sonatypePublishTo.value

// [If necessary] Settings for using custom Nexus repositories:
sonatypeCredentialHost := "s01.oss.sonatype.org"
sonatypeRepository := "https://s01.oss.sonatype.org/service/local"
```

### $HOME/.sbt/(sbt-version 0.13 or 1.0)/sonatype.sbt

For the authentication to Sonatype API, you need to set your Sonatype account information (user name and password) in the global sbt settings. To protect your password, never include this file within your project.

```scala
credentials += Credentials("Sonatype Nexus Repository Manager",
        "oss.sonatype.org",
        "(Sonatype user name)",
        "(Sonatype password)")
```

### (project root)/sonatype.sbt

sbt-sonatype is an auto-plugin, which will automatically configure your build. There are a few settings though that you need to define by yourself:

  * `sonatypeProfileName`
     * This is your Sonatype acount profile name, e.g. `org.xerial`. If you do not set this value, it will be the same with the `organization` value.
  * `pomExtra`
     * A fragment of Maven's pom.xml. You must define url, licenses, scm and developers tags in this XML to satisfy [Central Repository sync requirements](http://central.sonatype.org/pages/requirements.html).

Example settings:
```scala
// Your profile name of the sonatype account. The default is the same with the organization value
sonatypeProfileName := "(your organization. e.g., org.xerial)"

// To sync with Maven central, you need to supply the following information:
publishMavenStyle := true

// Open-source license of your choice
licenses := Seq("APL2" -> url("http://www.apache.org/licenses/LICENSE-2.0.txt"))

// Where is the source code hosted: GitHub or GitLab?
import xerial.sbt.Sonatype._
sonatypeProjectHosting := Some(GitHubHosting("username", "projectName", "user@example.com"))
// or
sonatypeProjectHosting := Some(GitLabHosting("username", "projectName", "user@example.com"))

// or if you want to set these fields manually
homepage := Some(url("https://(your project url)"))
scmInfo := Some(
  ScmInfo(
    url("https://github.com/(account)/(project)"),
    "scm:git@github.com:(account)/(project).git"
  )
)
developers := List(
  Developer(id="(your id)", name="(your name)", email="(your e-mail)", url=url("(your home page)"))
)
```

## Publishing Your Artifact

The basic steps for publishing your artifact to the Central Repository are as follows:

  * `publishSigned` to deploy your artifact to a local staging repository.
  * `sonatypeBundleRelease` (since sbt-sonatype 3.4)
    * This command is equivalent to running `; sonatypePrepare; sonatypeBundleUpload; sonatypeRelease`.
    * Internally `sonatypeRelease` will do `sonatypeClose` and `sonatypePromote` in one step.
      * `sonatypeClose` closes your staging repository at Sonatype. This step verifies Maven central sync requirement, GPG-signature, javadoc
   and source code presence, pom.xml settings, etc.
      * `sonatypePromote` command verifies the closed repository so that it can be synchronized with Maven central.

Note: If your project version has "SNAPSHOT" suffix, your project will be published to the [snapshot repository](http://oss.sonatype.org/content/repositories/snapshots) of Sonatype, and you cannot use `sonatypeBundleUpload` or `sonatypeRelease` command.

## Commands

Usually, we only need to run `sonatypeBundleRelease` command in sbt-sonatype:
* __sonatypeBundleRelease__
  * This will run a sequence of commands `; sonatypePrepare; sonatypeBundleUpload; sonatypeRelease` in one step.
  * You must run `publishSigned` before this command to create a local staging bundle.

### Individual Step Commands
* __sonatypePrepare__
  * Drop the exising staging repositories (if exist) and create a new staging repository using `sonatypeSessionName` as a unique key.
  * This will update `sonatypePublishTo` setting.
  * For cross-build projects, make sure running this command only once at the beginning of the release process.
    * Usually using sonatypeBundleUpload should be sufficient, but if you need to parallelize artifact uploads, run `sonatypeOpen` before each upload to reuse the already created stging repository.
* __sonatypeBundleUpload__
  * Upload your local staging folder contents to a remote Sonatype repository.
* __sonatypeOpen__
  * This command is necessary only when you need to parallelize `publishSigned` task. For small/medium-size projects, using only `sonatypePrepare` would work.
  * This opens the existing staging repository using `sonatypeSessionName` as a unique key. If it doesn't exist, create a new one. It will update`sonatypePublishTo`
* __sonatypeRelease__ (repositoryId)?
  * Close (if needed) and promote a staging repository. After this command, the uploaded artifacts will be synchronized to Maven central.

### Batch Operations
* __sonatypeDropAll__ (sonatypeProfileName)?
   * Drop all staging repositories.
* __sonatypeReleaseAll__ (sonatypeProfileName)?
  * Close and promote all staging repositories (Useful for cross-building projects)

## Other Commmands
* __sonatypeBundleClean__
  * Clean a local bundle folder
* __sonatypeClean__
  * Clean a remote staging repository which has `sonatypeSessionName` key.
* __sonatypeStagingProfiles__
  * Show the list of staging profiles, which include profileName information.
* __sonatypeLog__
  * Show the staging activity logs
* __sonatypeClose__
  * Close the open staging repository (= requirement verification)
* __sonatypePromote__
  * Promote the closed staging repository (= sync to Maven central)
* __sonatypeDrop__
  * Drop an open or closed staging repository

## Advanced Build Settings

### Sequential Upload Release (Use this for small projects)

```scala
> ; publishSigned; sonatypeBundleRelease
```

For cross-building projects, use `+ publishSigned`:
```scala
> ; + publishSigned; sonatypeBundleRelease
```
### Parallelizing Builds When Sharing A Working Folder

When you are sharing a working folder, you can parallelize publishSigned step for each module or for each Scala binary version:

- Run multiple publishSigned tasks in parallel
- Finally, run `sonatypeBundleRelease`

### Parallelizing Builds When Not Sharing Any Working Folder

If you are not sharing any working directory (e.g., Travis CI), to parallelize the release process, you need to publish a bundle for each build because Sonatype API only supports uploading one bundle per a staging repository.

Here is an example to parallelize your build for each Scala binary version:

ReleaseNotes.md
Release Notes
===

See [the release note page](https://github.com/xerial/sbt-sonatype/releases) for the latest release notes. 

# 3.9.14
 
- Upgraded airframe-http to 22.11.0
- Add more Sonatype API error handling
- Upgrade to sbt 1.7.3
- Use sbt-dynver for versioning

# 3.9.13

- Upgraded airframe-http to 22.5.0 to fix JDK17-specific error [#293](https://github.com/xerial/sbt-sonatype/issues/293)
- Reverted the fix for #276 due to the regression


# 3.9.12 
- DO NOT USE THIS VERSION. A regression, which failed to find a proper sonatype credential, is reported. Use 3.9.11 instead
- A fix for [#276](https://github.com/xerial/sbt-sonatype/issues/276): Always use `ThisBuild / sonatypeCredentialHost` setting [#285](https://github.com/xerial/sbt-sonatype/pull/285)

# 3.9.11
- Use unique names for resolvers to support s01.oss.sonatype.org [#279](https://github.com/xerial/sbt-sonatype/pull/#279)
- Increase the max retry to 100, timeoutMillis to 3 hours for big projects
- [internal] Upgrade to airframe 22.1.0

# 3.9.10
- Improve the cache management to support s01.oss.sonatype.org
- Internal: Upgrade to airframe 21.8.1 (no scala-parser-combinator dependency)

# 3.9.9
- Remove scala-parser-combinator 2.x dependency. This is a workaround for sbt-pgp, which still depends on scala-parser-combinators 1.x.

# 3.9.8
- Scope sonatypeCredentialHost to the build instead of projects. (#242)
- internal: Upgrade to airframe 21.8.0

# 3.9.7
- 2021-03-12
- Fixes a bug in releasing snapshot artifacts to the new s01.oss.sonatype.org repository.

# 3.9.6
- 2021-03-10
- Support a new sonatype host with `sonatypeCredentialHost := "s01.oss.sonatype.org"` setting
- internal: Upgrade to airframe 21.3.0

# 3.9.5
- 2020-11-02
- Fixes an issue when different versions of sub modules exist [#197](https://github.com/xerial/sbt-sonatype/issues/197)

# 3.9.4
- 2020-07-03
- Use URLConnection based HTTP client to remove the dependency to Finagle and Jackson
- internal: Upgrade to airframe-http 20.6.2

# 3.9.3
- 2020-06-19
- Fixes an issue when using `sbt "sonatypeReleaseAll (sonatypeProfileName)` standalone without build.sbt file. 
- Using GitHub Actions for CI

# 3.9.2
- 2020-03-27
- Use a longer total timeout (upto 10 min.) for individual HTTP requests to address an issue like [#149](https://github.com/xerial/sbt-sonatype/issues/149)

# 3.9.1
- 2020-03-27
- Adjust the wait interval for close/promote stages
- Fixed a bug so as not to retry the create staging repository request 

# 3.9.0
- 2020-03-26
- __This version has an issue in creating staging repositories. Use 3.9.1__
- Add sonatypeTimeoutMillis to wait longer time until close/promote stage completes. The default is 60 minutes.
- Changed the backend HTTP client to airframe-http-finagle
- Use JSON REST API of Sonatype instead of handling XML
- Add retry for each HTTP request and bundle upload
- Use ANSI colors for logging messages
- Improved error messages

# 3.8.1
- 2019-11-20
- Fix for the credential resolver to support sbt-gpg [#124](https://github.com/xerial/sbt-sonatype/pull/124)

# 3.8
- 2019-09-27
- Show log message periodically to avoid Travis CI 10 minutes timeout [#108](https://github.com/xerial/sbt-sonatype/issues/108)

# 3.7
- 2019-09-11
- Upgrade http-client version to address NoClassDefFound error.

# 3.6
- 2019-09-06
- Support dynamic versions generated by sbt-dynver.

# 3.5
- 2019-09-06
- A minor fix for multi-module projects to use a single staging folder [#94](https://github.com/xerial/sbt-sonatype/pull/94)

# 3.4
- 2019-09-05
- Added sonatypeBundleRelease as a short cut for `sonatypePrepare; sonatypeUploadBundle; sonatypeRelease`

# 3.2
- 2019-09-05
- Re-built for sbt 1.2.8

# 3.1
- September 5, 2019
- Support `sonatypeBundleUpload`, which makes uploading hundreds of artifacts much faster than directly uploading them to Sonatype. [#89](https://github.com/xerial/sbt-sonatype/issues/89)

# 3.0
- September 4, 2019
- Support building idempotent release flow using `sonatypePrepare`
- Support parallel artifact upload flows. See https://github.com/xerial/sbt-sonatype/#uploading-artifacts-in-parallel for the details.
- Parallelize the processing of sonatypeReleaseAll and sonatypeDropAll
- Drop `sontaypeList`. Use sonatypeStagingRepositoryProfiles instead
- From this version, sbt-sonatype no longer overwrites `publishTo` settings automatically after `sonatypeOpen` command. Make sure setting `publishTo := sonatypePublishTo.value`.

# 2.6
- Added `sonatypeDropAll`

# 2.5
- Fixes issue #79

# 2.4
- Fixes a bug in propagating publishTo setting to sub projects [#76](https://github.com/xerial/sbt-sonatype/issues/76)
- Drop sbt 0.13.x support

# 2.3
- Allow setting the credentials with the environment variables `SONATYPE_USERNAME` and `SONATYPE_PASSWORD`

# 2.2
- Fixes typo [#61](https://github.com/xerial/sbt-sonatype/pull/61) in GitHubHosting and GitLabHosting capitilization
- If you are not using these keys, no need to upgrade to this version.

# 2.1
- Fixes [#55](https://github.com/xerial/sbt-sonatype/issues/55) with `sonatypePublishTo` setting:
```scala
publishTo := sonatypePublishTo
```
- Add shortcut `sonatypeProjectHosting` for quickly setting `homepage`, `scmInfo` and `developers`


# 2.0
- Support sbt-0.13, 1.0.0-M5, 1.0.0-M6, and 1.0.0-RC3

# 2.0.0-M1
- Add sbt-1.0.0-M5 (for Scala 2.12), sbt-0.13.x support
- **IMPORTANT** sbt-sonatype no longer modifies `publishTo` settings. You need to manually set the Sonatype repository name as below:
```scala
publishTo := Some(
  if (isSnapshot.value)
    Opts.resolver.sonatypeSnapshots
  else
    Opts.resolver.sonatypeStaging
)
```

# 1.1
- Add sonatypeOpen command for supporting more complex workflows (see the usage example in [workflow.md](workflow.md))

# 1.0
- The first major release (stable)

# 0.5.1

src/main/scala/xerial/sbt/Sonatype.scala
//--------------------------------------
//
// Sonatype.scala
// Since: 2014/01/05
//
//--------------------------------------

package xerial.sbt

import sbt.Keys._
import sbt._
import sbt.librarymanagement.MavenRepository
import wvlet.log.{LogLevel, LogSupport}
import xerial.sbt.sonatype.SonatypeClient.StagingRepositoryProfile
import xerial.sbt.sonatype.SonatypeService._
import xerial.sbt.sonatype.{SonatypeClient, SonatypeException, SonatypeService}

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, ExecutionContext, Future}
import scala.util.hashing.MurmurHash3

/** Plugin for automating release processes at Sonatype Nexus
  */
object Sonatype extends AutoPlugin with LogSupport {
  wvlet.log.Logger.init

  trait SonatypeKeys {
    val sonatypeRepository  = settingKey[String]("Sonatype repository URL: e.g. https://oss.sonatype.org/service/local")
    val sonatypeProfileName = settingKey[String]("Profile name at Sonatype: e.g. org.xerial")
    val sonatypeCredentialHost          = settingKey[String]("Credential host. Default is oss.sonatype.org")
    val sonatypeDefaultResolver         = settingKey[Resolver]("Default Sonatype Resolver")
    val sonatypePublishTo               = settingKey[Option[Resolver]]("Default Sonatype publishTo target")
    val sonatypePublishToBundle         = settingKey[Option[Resolver]]("Default Sonatype publishTo target")
    val sonatypeTargetRepositoryProfile = settingKey[StagingRepositoryProfile]("Stating repository profile")
    val sonatypeProjectHosting =
      settingKey[Option[ProjectHosting]]("Shortcut to fill in required Maven Central information")
    val sonatypeSessionName      = settingKey[String]("Used for identifying a sonatype staging repository")
    val sonatypeTimeoutMillis    = settingKey[Int]("milliseconds before giving up Sonatype API requests")
    val sonatypeBundleClean      = taskKey[Unit]("Clean up the local bundle folder")
    val sonatypeBundleDirectory  = settingKey[File]("Directory to create a bundle")
    val sonatypeBundleRelease    = taskKey[String]("Release a bundle to Sonatype")
    val sonatypeLogLevel         = settingKey[String]("log level: trace, debug, info warn, error")
    val sonatypeSnapshotResolver = settingKey[Resolver]("Sonatype snapshot resolver")
    val sonatypeStagingResolver  = settingKey[Resolver]("Sonatype staging resolver")
  }

  object SonatypeKeys extends SonatypeKeys {}

  object autoImport extends SonatypeKeys {}

  override def trigger         = allRequirements
  override def projectSettings = sonatypeSettings
  override def buildSettings   = sonatypeBuildSettings

  import autoImport._
  import complete.DefaultParsers._

  private implicit val ec = ExecutionContext.global

  val sonatypeLegacy = "oss.sonatype.org"
  val sonatype01     = "s01.oss.sonatype.org"

  lazy val sonatypeBuildSettings = Seq[Def.Setting[_]](
    sonatypeCredentialHost := sonatypeLegacy
  )
  lazy val sonatypeSettings = Seq[Def.Setting[_]](
    sonatypeProfileName    := organization.value,
    sonatypeRepository     := s"https://${(ThisBuild / sonatypeCredentialHost).value}/service/local",
    sonatypeProjectHosting := None,
    publishMavenStyle      := true,
    pomIncludeRepository := { _ =>
      false
    },
    credentials ++= {
      val alreadyContainsSonatypeCredentials: Boolean = credentials.value.exists {
        case d: DirectCredentials => d.host == sonatypeCredentialHost.value
        case _                    => false
      }
      if (!alreadyContainsSonatypeCredentials) {
        val env = sys.env.get(_)
        (for {
          username <- env("SONATYPE_USERNAME")
          password <- env("SONATYPE_PASSWORD")
        } yield Credentials(
          "Sonatype Nexus Repository Manager",
          sonatypeCredentialHost.value,
          username,
          password
        )).toSeq
      } else Seq.empty
    },
    homepage := homepage.value.orElse(sonatypeProjectHosting.value.map(h => url(h.homepage))),
    scmInfo  := sonatypeProjectHosting.value.map(_.scmInfo).orElse(scmInfo.value),
    developers := {
      val derived = sonatypeProjectHosting.value.map(h => List(h.developer)).getOrElse(List.empty)
      if (developers.value.isEmpty) derived
      else developers.value
    },
    sonatypePublishTo := Some(sonatypeDefaultResolver.value),
    sonatypeBundleDirectory := {
      (ThisBuild / baseDirectory).value / "target" / "sonatype-staging" / s"${(ThisBuild / version).value}"
    },
    sonatypeBundleClean := {
      IO.delete(sonatypeBundleDirectory.value)
    },
    sonatypePublishToBundle := {
      if (version.value.endsWith("-SNAPSHOT")) {
        // Sonatype snapshot repositories have no support for bundle upload,
        // so use direct publishing to the snapshot repo.
        Some(sonatypeSnapshotResolver.value)
      } else {
        Some(Resolver.file("sonatype-local-bundle", sonatypeBundleDirectory.value))
      }
    },
    sonatypeSnapshotResolver := {
      MavenRepository(
        s"${sonatypeCredentialHost.value.replace('.', '-')}-snapshots",
        s"https://${sonatypeCredentialHost.value}/content/repositories/snapshots"
      )
    },
    sonatypeStagingResolver := {
      MavenRepository(
        s"${sonatypeCredentialHost.value.replace('.', '-')}-staging",
        s"https://${sonatypeCredentialHost.value}/service/local/staging/deploy/maven2"
      )
    },
    sonatypeDefaultResolver := {
      val profileM   = sonatypeTargetRepositoryProfile.?.value
      val repository = sonatypeRepository.value
      val staged = profileM.map { stagingRepoProfile =>
        s"${sonatypeCredentialHost.value.replace('.', '-')}-releases" at s"${repository}/${stagingRepoProfile.deployPath}"
      }
      staged.getOrElse(if (version.value.endsWith("-SNAPSHOT")) {
        sonatypeSnapshotResolver.value
      } else {
        sonatypeStagingResolver.value
      })
    },
    sonatypeTimeoutMillis := 60 * 60 * 1000, // 60 minutes
    sonatypeSessionName   := s"[sbt-sonatype] ${name.value} ${version.value}",
    sonatypeLogLevel      := "info",
    commands ++= Seq(
      sonatypeBundleRelease,
      sonatypeBundleUpload,
      sonatypePrepare,
      sonatypeOpen,
      sonatypeClose,
      sonatypePromote,
      sonatypeDrop,
      sonatypeRelease,
      sonatypeClean,
      sonatypeReleaseAll,
      sonatypeDropAll,
      sonatypeLog,
      sonatypeStagingRepositoryProfiles,
      sonatypeStagingProfiles
    )
  )

  private def prepare(state: State, rest: SonatypeService): StagingRepositoryProfile = {
    val extracted      = Project.extract(state)
    val descriptionKey = extracted.get(sonatypeSessionName)
    state.log.info(s"Preparing a new staging repository for ${descriptionKey}")
    // Drop a previous staging repository if exists
    val dropTask = Future.apply(rest.dropIfExistsByKey(descriptionKey))
    // Create a new one
    val createTask = Future.apply(rest.createStage(descriptionKey))
    // Run two tasks in parallel
    val merged                     = dropTask.zip(createTask)
    val (droppedRepo, createdRepo) = Await.result(merged, Duration.Inf)
    createdRepo
  }

  private val sonatypeBundleRelease =
    newCommand("sonatypeBundleRelease", "Upload a bundle in sonatypeBundleDirectory and release it at Sonatype") {
      state: State =>
        withSonatypeService(state) { rest =>
          val repo       = prepare(state, rest)
          val extracted  = Project.extract(state)
          val bundlePath = extracted.get(sonatypeBundleDirectory)
          rest.uploadBundle(bundlePath, repo.deployPath)
          rest.closeAndPromote(repo)
          updatePublishSettings(state, repo)
        }
    }

  private val sonatypeBundleUpload = newCommand("sonatypeBundleUpload", "Upload a bundle in sonatypeBundleDirectory") {
    state: State =>
      val extracted  = Project.extract(state)
      val bundlePath = extracted.get(sonatypeBundleDirectory)
      withSonatypeService(state) { rest =>
        val repo = extracted.getOpt(sonatypeTargetRepositoryProfile).getOrElse {
          val descriptionKey = extracted.get(sonatypeSessionName)
          rest.openOrCreateByKey(descriptionKey)
        }
        rest.uploadBundle(bundlePath, repo.deployPath)
        updatePublishSettings(state, repo)
      }
  }

  private val sonatypePrepare = newCommand(
    "sonatypePrepare",
    "Clean (if exists) and create a staging repository for releasing the current version, then update publishTo"
  ) { state: State =>
    withSonatypeService(state) { rest =>
      val repo = prepare(state, rest)
      updatePublishSettings(state, repo)
    }
  }

  private val sonatypeOpen = newCommand(
    "sonatypeOpen",
    "Open (or create if not exists) to a staging repository for the current version, then update publishTo"
  ) { state: State =>
    withSonatypeService(state) { rest =>
      // Re-open or create a staging repository
      val descriptionKey = Project.extract(state).get(sonatypeSessionName)
      val repo           = rest.openOrCreateByKey(descriptionKey)
      updatePublishSettings(state, repo)
    }
  }

  private def updatePublishSettings(state: State, repo: StagingRepositoryProfile): State = {
    val extracted = Project.extract(state)
    // accumulate changes for settings for current project and all aggregates
    state.log.info(s"Updating sonatypePublishTo settings...")
    val newSettings: Seq[Def.Setting[_]] = extracted.currentProject.referenced.flatMap { ref =>
      Seq(
        ref / sonatypeTargetRepositoryProfile := repo
      )
    } ++ Seq(
      sonatypeTargetRepositoryProfile := repo
    )

    val next = extracted.appendWithoutSession(newSettings, state)
    next
  }

  private val sonatypeClose = commandWithRepositoryId("sonatypeClose", "") { (state: State, arg: Option[String]) =>
    val extracted = Project.extract(state)
    val repoID    = arg.orElse(extracted.getOpt(sonatypeTargetRepositoryProfile).map(_.repositoryId))
    withSonatypeService(state) { rest =>
      val repo1 = rest.findTargetRepository(Close, repoID)
      val repo2 = rest.closeStage(repo1)
      extracted.appendWithoutSession(Seq(sonatypeTargetRepositoryProfile := repo2), state)
    }
  }

  private val sonatypePromote = commandWithRepositoryId("sonatypePromote", "Promote a staging repository") {
    (state: State, arg: Option[String]) =>
      val extracted = Project.extract(state)
      val repoID    = arg.orElse(extracted.getOpt(sonatypeTargetRepositoryProfile).map(_.repositoryId))
      withSonatypeService(state) { rest =>
        val repo1 = rest.findTargetRepository(Promote, repoID)
        val repo2 = rest.promoteStage(repo1)
        extracted.appendWithoutSession(Seq(sonatypeTargetRepositoryProfile := repo2), state)
      }
  }

  private val sonatypeDrop = commandWithRepositoryId("sonatypeDrop", "Drop a staging repository") {
    (state: State, arg: Option[String]) =>
      val extracted = Project.extract(state)
      val repoID    = arg.orElse(extracted.getOpt(sonatypeTargetRepositoryProfile).map(_.repositoryId))
      withSonatypeService(state) { rest =>
        val repo1 = rest.findTargetRepository(Drop, repoID)
        val repo2 = rest.dropStage(repo1)
        extracted.appendWithoutSession(Seq(sonatypeTargetRepositoryProfile := repo2), state)
      }
  }

  private val sonatypeRelease =
    commandWithRepositoryId("sonatypeRelease", "Publish with sonatypeClose and sonatypePromote") {
      (state: State, arg: Option[String]) =>
        val extracted = Project.extract(state)
        val repoID    = arg.orElse(extracted.getOpt(sonatypeTargetRepositoryProfile).map(_.repositoryId))
        withSonatypeService(state) { rest =>
          val repo1 = rest.findTargetRepository(CloseAndPromote, repoID)
          val repo2 = rest.closeAndPromote(repo1)
          extracted.appendWithoutSession(Seq(sonatypeTargetRepositoryProfile := repo2), state)
        }
    }

  private val sonatypeClean =
    newCommand("sonatypeClean", "Clean a staging repository for the current version if it exists") { state: State =>
      val extracted = Project.extract(state)
      withSonatypeService(state) { rest =>
        val descriptionKey = extracted.get(sonatypeSessionName)
        rest.dropIfExistsByKey(descriptionKey)
        state
      }
    }

  private val sonatypeReleaseAll =
    commandWithRepositoryId("sonatypeReleaseAll", "Publish all staging repositories to Maven central") {
      (state: State, arg: Option[String]) =>
        withSonatypeService(state, arg) { rest =>
          val tasks = rest.stagingRepositoryProfiles().map { repo =>
            Future.apply(rest.closeAndPromote(repo))
          }
          val merged = Future.sequence(tasks)
          Await.result(merged, Duration.Inf)
          state
        }
    }

  private val sonatypeDropAll = commandWithRepositoryId("sonatypeDropAll", "Drop all staging repositories") {
    (state: State, arg: Option[String]) =>
      withSonatypeService(state, arg) { rest =>
        val dropTasks = rest.stagingRepositoryProfiles().map { repo =>
          Future.apply(rest.dropStage(repo))
        }
        val merged = Future.sequence(dropTasks)
        Await.result(merged, Duration.Inf)
        state
      }
  }

  private val sonatypeLog = newCommand("sonatypeLog", "Show staging activity logs at Sonatype") { state: State =>
    withSonatypeService(state) { rest =>
      val alist = rest.activities
      if (alist.isEmpty)
        warn("No staging log is found")
      for ((repo, activities) <- alist) {
        info(s"Staging activities of $repo:")
        for (a <- activities) {
          a.showProgress
        }
      }
      state
    }
  }

  private val sonatypeStagingRepositoryProfiles =
    newCommand("sonatypeStagingRepositoryProfiles", "Show the list of staging repository profiles") { state: State =>
      withSonatypeService(state) { rest =>
        val repos = rest.stagingRepositoryProfiles()
        if (repos.isEmpty)
          warn(s"No staging repository is found for ${rest.profileName}")
        else {
          info(s"Staging repository profiles (sonatypeProfileName:${rest.profileName}):")
          info(repos.mkString("\n"))
        }
        state
      }
    }

  private val sonatypeStagingProfiles = newCommand("sonatypeStagingProfiles", "Show the list of staging profiles") {
    state: State =>
      withSonatypeService(state) { rest =>
        val profiles = rest.stagingProfiles
        if (profiles.isEmpty)
          warn(s"No staging profile is found for ${rest.profileName}")
        else {
          info(s"Staging profiles (sonatypeProfileName:${rest.profileName}):")
          info(profiles.mkString("\n"))
        }
        state
      }
  }

  case class ProjectHosting(
      domain: String,
      user: String,
      fullName: Option[String],
      email: String,
      repository: String
  ) {
    def homepage  = s"https://$domain/$user/$repository"
    def scmUrl    = s"git@$domain:$user/$repository.git"
    def scmInfo   = ScmInfo(url(homepage), scmUrl)
    def developer = Developer(user, fullName.getOrElse(user), email, url(s"https://$domain/$user"))
  }

  object GitHubHosting {
    private val domain                                         = "github.com"
    def apply(user: String, repository: String, email: String) = ProjectHosting(domain, user, None, email, repository)
    def apply(user: String, repository: String, fullName: String, email: String) =
      ProjectHosting(domain, user, Some(fullName), email, repository)
  }

  object GitLabHosting {
    private val domain                                         = "gitlab.com"
    def apply(user: String, repository: String, email: String) = ProjectHosting(domain, user, None, email, repository)
    def apply(user: String, repository: String, fullName: String, email: String) =
      ProjectHosting(domain, user, Some(fullName), email, repository)
  }

  // aliases
  @deprecated("Use GitHubHosting (capital H) instead", "2.2")
  val GithubHosting = GitHubHosting
  @deprecated("Use GitLabHosting (capital L) instead", "2.2")
  val GitlabHosting = GitLabHosting

  private val repositoryIdParser: complete.Parser[Option[String]] =
    (Space ~> token(StringBasic, "(sonatype staging repository id)")).?.!!!(
      "invalid input. please input a repository id"
    )

  private val sonatypeProfileParser: complete.Parser[Option[String]] =
    (Space ~> token(StringBasic, "(sonatypeProfileName)")).?.!!!(
      "invalid input. please input sonatypeProfileName (e.g., org.xerial)"
    )

  private def getCredentials(extracted: Extracted, state: State) = {
    val (nextState, credential) = extracted.runTask(credentials, state)
    credential
  }

  private def withSonatypeService(state: State, profileName: Option[String] = None)(
      body: SonatypeService => State
  ): State = {
    val extracted = Project.extract(state)
    val logLevel  = LogLevel(extracted.get(sonatypeLogLevel))
    wvlet.log.Logger.setDefaultLogLevel(logLevel)

    // #276 Need to use the root configuration
    val credentialHost         = extracted.get(ThisBuild / sonatypeCredentialHost)
    val currentSonatypeProfile = profileName.getOrElse(extracted.get(sonatypeProfileName))
    val repositoryUrl          = extracted.get(sonatypeRepository)
    val timeoutMillis          = extracted.get(sonatypeTimeoutMillis)

    val creds = getCredentials(extracted, state)
    val hashsum: String = {
      val input = Vector(repositoryUrl, creds.toString(), credentialHost).mkString("-")
      MurmurHash3.stringHash(input).abs.toString()
    }

    val sonatypeClient = new SonatypeClient(
      repositoryUrl = repositoryUrl,
      cred = creds,
      credentialHost = credentialHost,
      timeoutMillis = timeoutMillis
    )
    val service = new SonatypeService(
      sonatypeClient,
      currentSonatypeProfile,
      Some(hashsum)
    )
    try {
      body(service)
    } catch {
      case e: SonatypeException =>
        error(e.toString)
        state.fail
      case e: Throwable =>
        error(e)
        state.fail
    } finally {
      service.close()
    }
  }

  private def newCommand(name: String, briefHelp: String)(body: State => State) = {
    Command.command(name, briefHelp, briefHelp)(body)
  }

  private def commandWithRepositoryId(name: String, briefHelp: String) =
    Command(name, (name, briefHelp), briefHelp)(_ => repositoryIdParser)(_)

}
